### ISSUE TITLE
Regular axios POST to login works, vue-auth POST doesn't

### ISSUE THREAD

**Issue by sergiocastrovale** [websanova/vue-auth#221]
Basically what the title says. 

This is my query string (according to axios [that's how you send data in application/x-www-form-urlencoded format](https://github.com/mzabriskie/axios#using-applicationx-www-form-urlencoded-format):
    
        computed: {
            formData: function() {
                var params = new URLSearchParams();

                params.append('grant_type', 'password');
                params.append('client_id', 'agentAuthApp');
                params.append('username', this.username);
                params.append('password', this.password);

                return params;
            }
        },

When I try to post data directly with axios, everything works fine:
   
          axiosLogin: function() {
                axios.post('auth/login', this.formData).then((response) => {
                    console.log('This works!', response);
                }).catch((e) => {
                    console.log('Error', e);
                });
            },

However if I try to do so following vue-auth's example ... 

    
           login: function() {
                var redirect = this.$auth.redirect();
                
                this.$auth.login({
                    data: this.formData,
                    rememberMe: this.rememberMe,
                    redirect: { name: redirect ? redirect.from.name : 'reservations' },
                     success: function(result) {
                        console.log('Not working', result);
                    },
                    error: function(response) {
                        this.error = response.data;
                    }
                });
            },

... Or even with a JSON object instead of a query string ...

            login: function() {
                var redirect = this.$auth.redirect();

                this.$auth.login({
                    data: { username: this.username, password: this.password },
                    success: function(result) {
                        console.log('Not working', result);
                    },
                    error: function(response) {
                        this.error = response.data;
                    }
                });
            },

... I am getting a **_Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'http://localhost:8081' is therefore not allowed access._** error. Any idea why? I even tried to POST with jQuery just to be sure and it also works as expected.

One thing to note is that when I inspect the request, I'm getting an OPTIONS request method to the API, not a POST, which is weird. I tried to manually add interceptors like this:

                axios.interceptors.request.use((config) => {
                    config.headers.common['Content-Type'] = 'application/x-www-form-urlencoded';
                    return config;
                }, function (error) {
                    return Promise.reject(error);
                });

but the result was the same.

**Comment by websanova** [comment]
Seems like CORS issue with your server

**Comment by sergiocastrovale** [comment]
I'm pretty sure it isn't, because the basic axios request works, so all headers are set and the POST returns successfully. Do you have any idea on how to debug this?

**Comment by websanova** [comment]
I don't know what tested, but I can tell you the arguments in auth login
just get passed directly to axios. There is no magic there.

Can you verify the appropriate header is sent sure the options request?

**Comment by sergiocastrovale** [comment]
Hi, I've just fixed this. Not sure if this has something to do with a part of my server code or not, but the thing is the solution is aligned with what axios has on its documentation and _not_ what you have in your examples:

               this.$auth.login({
                    data: { username: 'abc', password: 'test', grant_type: 'password', clientId: 'agentAuthApp' }, // This DOESN'T work
                    data: "grant_type=password&username=abc&password=test&client_id=agentAuthApp" // This works
                    rememberMe: false,
                    fetchUser: false,
                    redirect: '/test',
                    success: function(response) {
                        console.log(response);
                    },
                    error: function(response) {
                        this.error = response.data;
                    }
                });

So, in short, when handling form-urlencoded POST requests with axios we cannot send a JSON object; we have to stringify it in order for it to work.

**Comment by websanova** [comment]
Ok so this is axios then

On Aug 9, 2017 19:27, "Sérgio Castro" <notifications@github.com> wrote:

Closed #221 <https://github.com/websanova/vue-auth/issues/221>.

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/websanova/vue-auth/issues/221#event-1198897577>, or mute
the thread

**Comment by sergiocastrovale** [comment]
I think it's a problem with the documentation of this package because if you follow what's recommended you'll be sending a JSON object to axios, while it needs to have a query string instead (in 'application/x-www-form-urlencoded' POSTs).

In short, you **can't** use axios and send this (example taken from the doc):

$auth.login({
    rememberMe: true,
    data: { email: 'test@example.com', password: 'abcd1234' }
})

But instead you need something like this:

$auth.login({
    rememberMe: true,
    data: "email=test@example.com&password=abcd1234"
})

**Comment by websanova** [comment]
I don't think so, but ok...

**Comment by sergiocastrovale** [comment]
Just thought it could be helpful for other people stumbling on this problem. You can test it out yourself easily as well, just copy the code samples above and you'll see what I mean.

Other than that, thanks a lot for the package, seems promising!

**Comment by dhax** [comment]
I can tell from using this lib, that I have no problem with axios and sending json. It works like documented, there must be something else wrong in your specific setup.
